### DISCUSSION TITLE
SSH client

### DISCUSSION THREAD

**Discussion by markus-li** [DustinBrett/daedalOS#117] [Feature Requests]
Have you considered adding an SSH client? I haven't tested any running 100% in the browser, the ones I've used involve the host and runs through a websocket. One direct SSH client I know of is FireSSH, but that is (was) Firefox specific. Now it doesn't even work in Firefox anymore.
Another is [SSHy](https://github.com/stuicey/SSHy), but it's not maintained it seems, maybe there's a fork somewhere being maintained?

**Comment by DustinBrett** [comment]
I think I gave it a brief thought but my main issue with the network connected apps is they require a connection that typically isn't possibly without a WebSocket proxy. I would like to do it and I will add this to my feature requests to do list. I just need to think of a way to avoid doing anything in the backend as my goal is client side only. I would be fine if it was possible to use something already public with CORS/WebSocket allowed. Thanks for the comment!

**Reply by markus-li** [reply]
I understand your desire to keep it 100% in-browser when possible, but if you would add a very simple websocket proxy you could use for any and all things network related which can't be done in a browser directly and keep the majority of the logic inside the browser maybe that would be a good enough of a compromise? When/if it is made possible to use a different way to terminate a connection through other means at least you have the browser side already there. 
It is a nice experience with SSH in the browser, I'm running Code inside DaedalOS and I'm really enjoying using the terminal there!

**Reply by DustinBrett** [reply]
Thanks for the feedback! I will indeed be looking into this more and getting SSH into the terminal so it can do direct connections. I actually do have a WebSocket proxy I use with v86 but it's not mine and often does not work. I am hopeful one day to have something like [Direct Sockets](https://github.com/WICG/direct-sockets/blob/main/docs/explainer.md) so I can connect from the browser without a middle-man. Another option could be SSH servers that also serve via WebSockets. Such is being used for my IRC Client where it can directly connect to servers support [IRCv3 WebSocket Spec](https://ircv3.net/specs/extensions/websocket). As for hosting something myself server side, I am trying at all costs to avoid that can of worms from a liability pov of being a proxy server for whatever.

**Reply by hello-smile6** [reply]
@DustinBrett Why not just use the [relay.widgetry.org](https://relay.widgetry.org/) websocket proxy and implement a DHCP client via WebAssembly?

**Reply by DustinBrett** [reply]
Well that sounds like a cool idea indeed. Any suggestions where to begin for something like that? That server is indeed the one I use for v86, but I haven't thought much about expanding it's use and definitely hadn't thought of a DHCP client or anything like that. Looking at the [code](https://github.com/tsl0922/ttyd/blob/main/html/src/components/terminal/index.tsx) for [ttyd](https://github.com/tsl0922/ttyd), I think at least for SSH, this could be an option I could add to my Terminal. But expanding WebSocket proxy use would also be cool. Then I'd need a systray for the networking info. :-)

**Reply by hello-smile6** [reply]
Not sure how, currently working on a TCP relay that will allow access to a whitelist of servers via websockets. Once I have a working server, I'll create a demo.

**Reply by hello-smile6** [reply]
@DustinBrett https://ws2tcp.9pfs.repl.co/demos/main.html
Sorry if the directory index is strange, I manually created it rather than using a package that would provide it.

**Reply by hello-smile6** [reply]
@DustinBrett Would it be possible to use SSH via a relay like that?
